File: Halachic Times/GeoNames/src/main/java/net/sf/geonames/Countries.java
/*
 * Source file of the Halachic Times project.
 * Copyright (c) 2012. All Rights Reserved.
 * 
 * The contents of this file are subject to the Mozilla Public License Version
 * 2.0 (the "License"); you may not use this file except in compliance with the
 * License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/2.0
 *
 * Contributors can be contacted by electronic mail via the project Web pages:
 * 
 * http://sourceforge.net/projects/halachictimes
 * 
 * http://halachictimes.sourceforge.net
 *
 * Contributor(s):
 *   Moshe Waisberg
 * 
 */
package net.sf.geonames;

import com.sun.org.apache.xml.internal.serializer.OutputPropertiesFactory;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Result;
import javax.xml.transform.Source;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

/**
 * Countries.
 *
 * @author Moshe Waisberg
 */
public class Countries extends Cities {

    /** The number of main vertices per region border. */
    private static final int VERTICES_COUNT = 16;

    /**
     * Constructs a new countries.
     */
    public Countries() {
        super();
    }

    public static void main(String[] args) throws Exception {
        File countryInfoFile = new File("GeoNames/res/countryInfo.txt");
        File shapesFile = new File("GeoNames/res/shapes_simplified_low.txt");

        Countries countries = new Countries();
        Collection<CountryInfo> names = countries.loadInfo(countryInfoFile);
        Collection<GeoShape> shapes = countries.loadShapes(shapesFile);
        Collection<CountryRegion> regions = countries.toRegions(names, shapes);

        countries.writeAndroidXML(regions);
    }

    /**
     * Transform the list of names to a list of country regions.
     *
     * @param names
     *         the list of places.
     * @return the list of regions.
     */
    public Collection<CountryRegion> toRegions(Collection<GeoName> names) {
        Map<String, CountryRegion> regions = new TreeMap<String, CountryRegion>();
        String countryCode;
        CountryRegion region;

        for (GeoName name : names) {
            countryCode = name.getCountryCode();

            if (!regions.containsKey(countryCode)) {
                region = new CountryRegion(countryCode);
                regions.put(countryCode, region);
            } else
                region = regions.get(countryCode);
            region.addLocation(name.getLatitude(), name.getLongitude());
        }

        return regions.values();
    }

    /**
     * Write the list of names as arrays in Android resource file format.
     *
     * @param countries
     *         the list of countries.
     * @throws ParserConfigurationException
     *         if a DOM error occurs.
     * @throws TransformerException
     *         if a DOM error occurs.
     */
    public void writeAndroidXML(Collection<CountryRegion> countries) throws ParserConfigurationException, TransformerException {
        List<CountryRegion> sorted = null;
        if (countries instanceof List)
            sorted = (List<CountryRegion>) countries;
        else
            sorted = new ArrayList<CountryRegion>(countries);
        Collections.sort(sorted, new RegionComparator());

        DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = builderFactory.newDocumentBuilder();
        Document doc = builder.newDocument();

        Element resources = doc.createElement(ANDROID_ELEMENT_RESOURCES);
        resources.appendChild(doc.createComment(HEADER));
        doc.appendChild(resources);

        Element countriesElement = doc.createElement(ANDROID_ELEMENT_STRING_ARRAY);
        countriesElement.setAttribute(ANDROID_ATTRIBUTE_NAME, "countries");
        countriesElement.setAttribute(ANDROID_ATTRIBUTE_TRANSLATABLE, "false");
        resources.appendChild(countriesElement);
        Element verticesCountElement = doc.createElement(ANDROID_ELEMENT_INTEGER_ARRAY);
        verticesCountElement.setAttribute(ANDROID_ATTRIBUTE_NAME, "vertices_count");
        verticesCountElement.setAttribute(ANDROID_ATTRIBUTE_TRANSLATABLE, "false");
        resources.appendChild(verticesCountElement);
        Element latitudesElement = doc.createElement(ANDROID_ELEMENT_INTEGER_ARRAY);
        latitudesElement.setAttribute(ANDROID_ATTRIBUTE_NAME, "latitudes");
        latitudesElement.setAttribute(ANDROID_ATTRIBUTE_TRANSLATABLE, "false");
        resources.appendChild(latitudesElement);
        Element longitudesElement = doc.createElement(ANDROID_ELEMENT_INTEGER_ARRAY);
        longitudesElement.setAttribute(ANDROID_ATTRIBUTE_NAME, "longitudes");
        longitudesElement.setAttribute(ANDROID_ATTRIBUTE_TRANSLATABLE, "false");
        resources.appendChild(longitudesElement);

        Element country, latitude, longitude, verticesCount;
        int[] pointIndexes;
        int pointIndex;
        int pointCount = 0;

        for (CountryRegion region : sorted) {
            pointIndexes = region.findMainVertices(VERTICES_COUNT);

            country = doc.createElement(ANDROID_ELEMENT_ITEM);
            country.setTextContent(region.getCountryCode());
            countriesElement.appendChild(country);

            pointCount = 0;
            for (int i = 0; i < VERTICES_COUNT; i++) {
                pointIndex = pointIndexes[i];

                if (pointIndex < 0)
                    break;

                latitude = doc.createElement(ANDROID_ELEMENT_ITEM);
                latitude.setTextContent(Integer.toString(region.ypoints[pointIndex]));
                latitudesElement.appendChild(latitude);
                longitude = doc.createElement(ANDROID_ELEMENT_ITEM);
                longitude.setTextContent(Integer.toString(region.xpoints[pointIndex]));
                longitudesElement.appendChild(longitude);
                pointCount++;
            }

            verticesCount = doc.createElement(ANDROID_ELEMENT_ITEM);
            verticesCount.setTextContent(Integer.toString(pointCount));
            verticesCountElement.appendChild(verticesCount);

        }

        File file = new File(getModulePath(), "values/countries.xml");
        file.getParentFile().mkdirs();

        Source src = new DOMSource(doc);
        Result result = new StreamResult(file);
        TransformerFactory xformerFactory = TransformerFactory.newInstance();
        Transformer transformer = xformerFactory.newTransformer();
        transformer.setOutputProperty(OutputKeys.INDENT, "yes");
        transformer.setOutputProperty(OutputPropertiesFactory.S_KEY_INDENT_AMOUNT, "4");
        transformer.transform(src, result);
    }

    /**
     * Load the list of countries.
     *
     * @param file
     *         the country info file.
     * @return the list of records.
     * @throws IOException
     *         if an I/O error occurs.
     */
    public Collection<CountryInfo> loadInfo(File file) throws IOException {
        return geoNames.parseCountries(file);
    }

    /**
     * Load the list of shapes.
     *
     * @param file
     *         the shapes file.
     * @return the list of records.
     * @throws IOException
     *         if an I/O error occurs.
     */
    public Collection<GeoShape> loadShapes(File file) throws IOException {
        return geoNames.parseShapes(file);
    }

    /**
     * Transform the list of countries to a list of country regions.
     *
     * @param names
     *         the list of countries.
     * @param shapes
     *         the list of country shapes.
     * @return the list of regions.
     */
    public Collection<CountryRegion> toRegions(Collection<CountryInfo> names, Collection<GeoShape> shapes) throws IOException {
        List<CountryRegion> regions = new ArrayList<>(names.size());
        Long geonameId;
        CountryRegion region;
        Map<Long, GeoShape> shapesById = new HashMap<>();
        GeoShape shape;

        for (GeoShape s : shapes) {
            shapesById.put(s.getGeoNameId(), s);
        }

        for (CountryInfo name : names) {
            geonameId = name.getGeoNameId();
            shape = shapesById.get(geonameId);
            if (shape != null) {
                region = CountryRegion.toRegion(name.getIso(), shape);
                if (region != null) {
                    regions.add(region);
                }
            }
        }

        return regions;
    }
}

File: Halachic Times/GeoNames/src/main/java/com/github/geonames/Countries.kt
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.geonames

import com.github.geonames.CountryRegion.Companion.VERTICES_COUNT
import com.github.geonames.CountryRegion.Companion.toRegion
import com.github.geonames.dump.NameShapesLow
import com.github.geonames.dump.PathCountryInfo
import com.github.geonames.dump.PathShapesLow
import java.io.File
import java.io.IOException
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException
import javax.xml.transform.OutputKeys
import javax.xml.transform.Result
import javax.xml.transform.Source
import javax.xml.transform.TransformerException
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult
import org.geonames.GeoNameId
import org.w3c.dom.Element

/**
 * Countries.
 *
 * @author Moshe Waisberg
 */
class Countries : Cities() {
    /**
     * Write the list of names as arrays in Android resource file format.
     *
     * @param countries the list of countries.
     * @throws ParserConfigurationException if a DOM error occurs.
     * @throws TransformerException if a DOM error occurs.
     */
    @Throws(ParserConfigurationException::class, TransformerException::class)
    fun writeAndroidXML(countries: Collection<CountryRegion>) {
        val sorted: List<CountryRegion> = countries.toList().sortedWith(RegionComparator())

        val builderFactory = DocumentBuilderFactory.newInstance()
        val builder = builderFactory.newDocumentBuilder()
        val doc = builder.newDocument()

        val resources = doc.createElement(ANDROID_ELEMENT_RESOURCES)
        resources.appendChild(doc.createComment(HEADER))
        doc.appendChild(resources)

        val countriesElement = doc.createElement(ANDROID_ELEMENT_STRING_ARRAY)
        countriesElement.setAttribute(ANDROID_ATTRIBUTE_NAME, "countries")
        countriesElement.setAttribute(ANDROID_ATTRIBUTE_TRANSLATABLE, "false")
        resources.appendChild(countriesElement)

        val verticesCountElement = doc.createElement(ANDROID_ELEMENT_INTEGER_ARRAY)
        verticesCountElement.setAttribute(ANDROID_ATTRIBUTE_NAME, "vertices_count")
        verticesCountElement.setAttribute(ANDROID_ATTRIBUTE_TRANSLATABLE, "false")
        resources.appendChild(verticesCountElement)

        val latitudesElement = doc.createElement(ANDROID_ELEMENT_INTEGER_ARRAY)
        latitudesElement.setAttribute(ANDROID_ATTRIBUTE_NAME, "latitudes")
        latitudesElement.setAttribute(ANDROID_ATTRIBUTE_TRANSLATABLE, "false")
        resources.appendChild(latitudesElement)

        val longitudesElement = doc.createElement(ANDROID_ELEMENT_INTEGER_ARRAY)
        longitudesElement.setAttribute(ANDROID_ATTRIBUTE_NAME, "longitudes")
        longitudesElement.setAttribute(ANDROID_ATTRIBUTE_TRANSLATABLE, "false")
        resources.appendChild(longitudesElement)

        var country: Element
        var latitude: Element
        var longitude: Element
        var verticesCount: Element
        var pointIndexes: IntArray
        var pointIndex: Int
        var pointCount: Int

        for (region in sorted) {
            pointIndexes = region.findMainVertices(region.boundary, VERTICES_COUNT)

            country = doc.createElement(ANDROID_ELEMENT_ITEM)
            country.textContent = region.countryCode
            countriesElement.appendChild(country)

            pointCount = 0
            for (i in 0 until VERTICES_COUNT) {
                pointIndex = pointIndexes[i]
                if (pointIndex < 0) continue

                latitude = doc.createElement(ANDROID_ELEMENT_ITEM)
                latitude.textContent = region.boundary.ypoints[pointIndex].toString()
                latitudesElement.appendChild(latitude)

                longitude = doc.createElement(ANDROID_ELEMENT_ITEM)
                longitude.textContent = region.boundary.xpoints[pointIndex].toString()
                longitudesElement.appendChild(longitude)

                pointCount++
            }
            verticesCount = doc.createElement(ANDROID_ELEMENT_ITEM)
            verticesCount.textContent = pointCount.toString()
            verticesCountElement.appendChild(verticesCount)
        }

        val file = File(modulePath, "values/countries.xml")
        file.parentFile.mkdirs()

        val src: Source = DOMSource(doc)
        val result: Result = StreamResult(file)
        val transformerFactory = TransformerFactory.newInstance()
        val transformer = transformerFactory.newTransformer()
        transformer.setOutputProperty(OutputKeys.INDENT, "yes")
        transformer.setOutputProperty(S_KEY_INDENT_AMOUNT, "4")
        transformer.transform(src, result)
    }

    /**
     * Load the list of countries.
     *
     * @param file the country info file.
     * @return the list of records.
     * @throws IOException if an I/O error occurs.
     */
    @Throws(IOException::class)
    fun loadInfo(file: File): Collection<CountryInfo> {
        return geoNames.parseCountries(file)
    }

    /**
     * Load the list of shapes.
     *
     * @param file the shapes file.
     * @param zippedName the zipped file name.
     * @return the list of records.
     * @throws IOException if an I/O error occurs.
     */
    @Throws(IOException::class)
    fun loadShapes(file: File, zippedName: String? = null): Collection<GeoShape> {
        return geoNames.parseShapes(file, zippedName)
    }

    /**
     * Transform the list of countries to a list of country regions.
     *
     * @param names the list of countries.
     * @param shapes the list of country shapes.
     * @return the list of regions.
     */
    @Throws(IOException::class)
    fun toRegions(
        names: Collection<CountryInfo>,
        shapes: Collection<GeoShape>
    ): Collection<CountryRegion> {
        val regions = mutableListOf<CountryRegion>()
        var geoNameId: GeoNameId
        val shapesById = shapes.associateBy { it.geoNameId }
        var countryCode: String
        var shape: GeoShape
        for (name in names) {
            geoNameId = name.geoNameId
            countryCode = name.iso ?: continue
            shape = shapesById[geoNameId] ?: continue
            regions.add(toRegion(countryCode, shape))
        }
        return regions
    }

    companion object {
        @Throws(Exception::class)
        @JvmStatic
        fun main(args: Array<String>) {
            var countryInfoPath = PathCountryInfo
            var shapesPath = PathShapesLow
            if (args.isNotEmpty()) {
                countryInfoPath = args[0]
                if (args.size > 1) {
                    shapesPath = args[1]
                }
            }
            val countries = Countries()
            val names = countries.loadInfo(File(countryInfoPath))
            val shapes = countries.loadShapes(File(shapesPath), NameShapesLow)
            val regions = countries.toRegions(names, shapes)
            countries.writeAndroidXML(regions)
        }
    }
}
